File: inter_zig/src/lexer.zig
const std = @import("std");
const token = @import("token.zig");

const expect = std.testing.expect;

pub fn new(input: []const u8) Lexer {
    var l = Lexer{
        .input = input,
        .position = 0,
        .readPosition = 0,
        .ch = undefined,
    };
    l.readChar();
    return l;
}

pub const Lexer = struct {
    input: []const u8,
    position: u64,
    readPosition: u64,
    ch: u8,

    pub fn readChar(self: *Lexer) void {
        if (self.readPosition >= self.input.len) {
            self.ch = 0;
        } else {
            self.ch = self.input[self.readPosition];
        }

        self.position = self.readPosition;
        self.readPosition += 1;
    }

    //invalid
    pub fn nextToken2(self: *Lexer) token.Token {
        var tok: token.Token = undefined;

        self.skipWhitespace();

        switch (self.ch) {
            '=' => {
                if (self.peekChar() == '=') {
                    self.readChar();
                    tok = token.Token{ .type = token.TokenTypes.EQ, .literal = "==" };
                } else {
                    tok = newToken(token.TokenTypes.ASSIGN, self.ch);
                }
            },
            ';' => {
                tok = newToken(token.TokenTypes.SEMICOLON, self.ch);
            },
            '(' => {
                tok = newToken(token.TokenTypes.LPAREN, self.ch);
            },
            ')' => {
                tok = newToken(token.TokenTypes.RPAREN, self.ch);
            },
            ',' => {
                tok = newToken(token.TokenTypes.COMMA, self.ch);
            },
            '+' => {
                tok = newToken(token.TokenTypes.PLUS, self.ch);
            },
            '{' => {
                tok = newToken(token.TokenTypes.LBRACE, self.ch);
            },
            '}' => {
                tok = newToken(token.TokenTypes.RBRACE, self.ch);
            },
            '-' => {
                tok = newToken(token.TokenTypes.MINUS, self.ch);
            },
            '!' => {
                if (self.peekChar() == '=') {
                    self.readChar();
                    tok = token.Token{ .type = token.TokenTypes.NOT_EQ, .literal = "!=" };
                } else {
                    tok = newToken(token.TokenTypes.BANG, self.ch);
                }
            },
            '/' => {
                tok = newToken(token.TokenTypes.SLASH, self.ch);
            },
            '*' => {
                tok = newToken(token.TokenTypes.ASTERISK, self.ch);
            },
            '<' => {
                tok = newToken(token.TokenTypes.LT, self.ch);
            },
            '>' => {
                tok = newToken(token.TokenTypes.GT, self.ch);
            },
            0 => {
                tok = newToken(token.TokenTypes.EOF, self.ch);
                tok.literal = "";
            },
            else => {
                if (isLetter(self.ch)) {
                    tok.literal = self.readIdentifier();
                    tok.type = token.lookupIdent(tok.literal);
                    return tok;
                } else if (isDigit(self.ch)) {
                    tok.type = token.TokenTypes.INT;
                    tok.literal = self.readNumber();
                    return tok;
                } else {
                    tok = newToken(token.TokenTypes.ILLEGAL, self.ch);
                }
            },
        }

        self.readChar();
        return tok;
    }

    pub fn nextToken(self: *Lexer) token.Token {
        var tok: token.Token = undefined;

        self.skipWhitespace();

        switch (self.ch) {
            '=' => {
                if (self.peekChar() == '=') {
                    const start = self.position;
                    self.readChar();
                    tok = token.Token{
                        .type = token.TokenTypes.EQ,
                        .literal = self.input[start .. self.position + 1],
                    };
                } else {
                    tok = token.Token{
                        .type = token.TokenTypes.ASSIGN,
                        .literal = self.input[self.position .. self.position + 1],
                    };
                }
            },
            ';' => {
                tok = token.Token{ .type = token.TokenTypes.SEMICOLON, .literal = self.input[self.position .. self.position + 1] };
            },
            '(' => {
                tok = token.Token{ .type = token.TokenTypes.LPAREN, .literal = self.input[self.position .. self.position + 1] };
            },
            ')' => {
                tok = token.Token{ .type = token.TokenTypes.RPAREN, .literal = self.input[self.position .. self.position + 1] };
            },
            ',' => {
                tok = token.Token{ .type = token.TokenTypes.COMMA, .literal = self.input[self.position .. self.position + 1] };
            },
            '+' => {
                tok = token.Token{ .type = token.TokenTypes.PLUS, .literal = self.input[self.position .. self.position + 1] };
            },
            '{' => {
                tok = token.Token{ .type = token.TokenTypes.LBRACE, .literal = self.input[self.position .. self.position + 1] };
            },
            '}' => {
                tok = token.Token{ .type = token.TokenTypes.RBRACE, .literal = self.input[self.position .. self.position + 1] };
            },
            '-' => {
                tok = token.Token{ .type = token.TokenTypes.MINUS, .literal = self.input[self.position .. self.position + 1] };
            },
            '!' => {
                if (self.peekChar() == '=') {
                    const start = self.position;
                    self.readChar();
                    tok = token.Token{ .type = token.TokenTypes.NOT_EQ, .literal = self.input[start .. self.position + 1] };
                } else {
                    tok = token.Token{ .type = token.TokenTypes.BANG, .literal = self.input[self.position .. self.position + 1] };
                }
            },
            '/' => {
                tok = token.Token{ .type = token.TokenTypes.SLASH, .literal = self.input[self.position .. self.position + 1] };
            },
            '*' => {
                tok = token.Token{ .type = token.TokenTypes.ASTERISK, .literal = self.input[self.position .. self.position + 1] };
            },
            '<' => {
                tok = token.Token{ .type = token.TokenTypes.LT, .literal = self.input[self.position .. self.position + 1] };
            },
            '>' => {
                tok = token.Token{ .type = token.TokenTypes.GT, .literal = self.input[self.position .. self.position + 1] };
            },
            '"' => {
                tok = token.Token{ .type = token.TokenTypes.STRING, .literal = self.readString() };
            },
            0 => {
                tok = token.Token{
                    .type = token.TokenTypes.EOF,
                    .literal = "",
                };
            },
            else => {
                if (isLetter(self.ch)) {
                    tok.literal = self.readIdentifier();
                    tok.type = token.lookupIdent(tok.literal);
                    return tok;
                } else if (isDigit(self.ch)) {
                    tok.type = token.TokenTypes.INT;
                    tok.literal = self.readNumber();
                    return tok;
                } else {
                    tok = token.Token{
                        .type = token.TokenTypes.ILLEGAL,
                        .literal = self.input[self.position .. self.position + 1],
                    };
                }
            },
        }

        self.readChar();
        return tok;
    }

    fn readIdentifier(self: *Lexer) []const u8 {
        const position = self.position;

        while (isLetter(self.ch)) {
            self.readChar();
        }

        return self.input[position..self.position];
    }

    //unused
    fn newToken(tokenType: token.TokenTypes, ch: u8) token.Token {
        return token.Token{
            .type = tokenType,
            .literal = &[_]u8{ch}, // salah di stack
        };
    }

    fn readNumber(self: *Lexer) []const u8 {
        const position = self.position;

        while (isDigit(self.ch)) {
            self.readChar();
        }

        return self.input[position..self.position];
    }
    fn skipWhitespace(self: *Lexer) void {
        while (self.ch == ' ' or self.ch == '\t' or self.ch == '\n' or self.ch == '\r') {
            self.readChar();
        }
    }

    fn peekChar(self: *Lexer) u8 {
        if (self.readPosition >= self.input.len) {
            return 0;
        } else {
            return self.input[self.readPosition];
        }
    }

    fn readString(self: *Lexer) []const u8 {
        const position = self.position + 1;

        while (true) {
            self.readChar();

            if (self.ch == '"' or self.ch == 0) {
                break;
            }
        }

        return self.input[position..self.position];
    }
};

fn isLetter(ch: u8) bool {
    //return 'a' <= ch and ch <= 'z' or 'A' <= ch and ch <= 'Z' or ch == '_';
    return std.ascii.isAlphabetic(ch) or ch == '_';
}

fn isDigit(ch: u8) bool {
    return std.ascii.isDigit(ch);
}
const TestNext = struct {
    expectedType: token.TokenTypes,
    expectedLiteral: []const u8,
};

test "next token" {
    const input =
        \\let five = 5;
        \\let ten = 10;
        \\let add = fn(x, y) {
        \\ x + y;
        \\};
        \\ let result = add(five, ten);
        \\!-/*5;
        \\5 < 10 > 5;
        \\ if (5 < 10) {
        \\  return true;
        \\ } else {
        \\  return false;
        \\ }
        \\
        \\ 10 == 10;
        \\ 10 != 9;
        \\ "foobar"
        \\ "foo bar"
    ;

    const tests = [_]TestNext{
        .{ .expectedType = token.TokenTypes.LET, .expectedLiteral = "let" },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "five" },
        .{ .expectedType = token.TokenTypes.ASSIGN, .expectedLiteral = "=" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "5" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        //
        .{ .expectedType = token.TokenTypes.LET, .expectedLiteral = "let" },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "ten" },
        .{ .expectedType = token.TokenTypes.ASSIGN, .expectedLiteral = "=" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "10" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        //
        .{ .expectedType = token.TokenTypes.LET, .expectedLiteral = "let" },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "add" },
        .{ .expectedType = token.TokenTypes.ASSIGN, .expectedLiteral = "=" },
        .{ .expectedType = token.TokenTypes.FUNCTION, .expectedLiteral = "fn" },
        .{ .expectedType = token.TokenTypes.LPAREN, .expectedLiteral = "(" },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "x" },
        .{ .expectedType = token.TokenTypes.COMMA, .expectedLiteral = "," },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "y" },
        .{ .expectedType = token.TokenTypes.RPAREN, .expectedLiteral = ")" },
        .{ .expectedType = token.TokenTypes.LBRACE, .expectedLiteral = "{" },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "x" },
        .{ .expectedType = token.TokenTypes.PLUS, .expectedLiteral = "+" },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "y" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        .{ .expectedType = token.TokenTypes.RBRACE, .expectedLiteral = "}" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        //
        .{ .expectedType = token.TokenTypes.LET, .expectedLiteral = "let" },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "result" },
        .{ .expectedType = token.TokenTypes.ASSIGN, .expectedLiteral = "=" },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "add" },
        .{ .expectedType = token.TokenTypes.LPAREN, .expectedLiteral = "(" },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "five" },
        .{ .expectedType = token.TokenTypes.COMMA, .expectedLiteral = "," },
        .{ .expectedType = token.TokenTypes.IDENT, .expectedLiteral = "ten" },
        .{ .expectedType = token.TokenTypes.RPAREN, .expectedLiteral = ")" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        .{ .expectedType = token.TokenTypes.BANG, .expectedLiteral = "!" },
        .{ .expectedType = token.TokenTypes.MINUS, .expectedLiteral = "-" },
        .{ .expectedType = token.TokenTypes.SLASH, .expectedLiteral = "/" },
        .{ .expectedType = token.TokenTypes.ASTERISK, .expectedLiteral = "*" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "5" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "5" },
        .{ .expectedType = token.TokenTypes.LT, .expectedLiteral = "<" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "10" },
        .{ .expectedType = token.TokenTypes.GT, .expectedLiteral = ">" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "5" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        .{ .expectedType = token.TokenTypes.IF, .expectedLiteral = "if" },
        .{ .expectedType = token.TokenTypes.LPAREN, .expectedLiteral = "(" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "5" },
        .{ .expectedType = token.TokenTypes.LT, .expectedLiteral = "<" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "10" },
        .{ .expectedType = token.TokenTypes.RPAREN, .expectedLiteral = ")" },
        .{ .expectedType = token.TokenTypes.LBRACE, .expectedLiteral = "{" },
        .{ .expectedType = token.TokenTypes.RETURN, .expectedLiteral = "return" },
        .{ .expectedType = token.TokenTypes.TRUE, .expectedLiteral = "true" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        .{ .expectedType = token.TokenTypes.RBRACE, .expectedLiteral = "}" },
        .{ .expectedType = token.TokenTypes.ELSE, .expectedLiteral = "else" },
        .{ .expectedType = token.TokenTypes.LBRACE, .expectedLiteral = "{" },
        .{ .expectedType = token.TokenTypes.RETURN, .expectedLiteral = "return" },
        .{ .expectedType = token.TokenTypes.FALSE, .expectedLiteral = "false" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        .{ .expectedType = token.TokenTypes.RBRACE, .expectedLiteral = "}" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "10" },
        .{ .expectedType = token.TokenTypes.EQ, .expectedLiteral = "==" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "10" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "10" },
        .{ .expectedType = token.TokenTypes.NOT_EQ, .expectedLiteral = "!=" },
        .{ .expectedType = token.TokenTypes.INT, .expectedLiteral = "9" },
        .{ .expectedType = token.TokenTypes.SEMICOLON, .expectedLiteral = ";" },
        .{ .expectedType = token.TokenTypes.STRING, .expectedLiteral = "foobar" },
        .{ .expectedType = token.TokenTypes.STRING, .expectedLiteral = "foo bar" },
    };

    var l = new(input);

    for (tests) |tt| {
        const tok = l.nextToken();
        if (tok.type != tt.expectedType) {
            std.debug.print("tokentype wrong. expected={s}, got={s} \n", .{ @tagName(tt.expectedType), @tagName(tok.type) });
        }
        if (!std.mem.eql(u8, tok.literal, tt.expectedLiteral)) {
            std.debug.print("literal wrong. expected={s}, got={s} \n", .{ tt.expectedLiteral, tok.literal });
        }
    }
}

File: inter_rust/src/lexer.rs
use std::hash::Hash;
use std::mem::discriminant;

#[derive(Debug, Eq, Clone)]
pub enum Token {
    ILLEGAL,
    EOF,
    IDENT(String),
    INT(Integer),

    COMMA,
    SEMICOLON,
    LPAREN,
    RPAREN,
    LBRACE,
    RBRACE,
    FUNCTION,
    LET,
    IF,
    TRUE,
    FALSE,
    ELSE,
    RETURN,

    //operators
    ASSIGN,
    PLUS,
    MINUS,
    BANG,
    ASTERISK,
    SLASH,
    LT,
    GT,
    NOT_EQ,
    EQ,
}

impl PartialEq for Token {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Token::IDENT(_), Token::IDENT(_)) => true,
            (Token::INT(a), Token::INT(b)) => a == b,
            _ => discriminant(self) == discriminant(other),
        }
    }
}

impl Hash for Token {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        discriminant(self).hash(state);
        if let Token::INT(i) = self {
            i.hash(state);
        }
    }
}

impl Token {
    pub fn literal(&self) -> String {
        match &self {
            &Self::ILLEGAL => "ILLEGAL".to_string(),
            &Self::LPAREN => "(".to_string(),
            &Self::RPAREN => ")".to_string(),
            &Self::LBRACE => "{".to_string(),
            &Self::RBRACE => "}".to_string(),
            &Self::EOF => "EOF".to_string(),
            &Self::IDENT(val) => val.to_string(),
            &Self::ASSIGN => "=".to_string(),
            &Self::PLUS => "+".to_string(),
            &Self::COMMA => ",".to_string(),
            &Self::SEMICOLON => ";".to_string(),
            &Self::FUNCTION => "FUNCTION".to_string(),
            &Self::LET => "LET".to_string(),
            &Self::INT(val) => val.get_string_value(),
            &Self::IF => String::from("if"),
            &Self::MINUS => "-".to_string(),
            &Self::BANG => "!".to_string(),
            &Self::ASTERISK => "*".to_string(),
            &Self::SLASH => "/".to_string(),
            &Self::LT => "<".to_string(),
            &Self::GT => ">".to_string(),
            &Self::TRUE => "true".to_string(),
            &Self::FALSE => "false".to_string(),
            &Self::ELSE => "else".to_string(),
            &Self::NOT_EQ => "!=".to_string(),
            &Self::EQ => "!=".to_string(),
            _ => "".to_string(),
        }
    }
}

#[derive(Hash, Clone, PartialEq, Eq, Debug)]
enum Integer {
    Integer32(i32),
    Integer64(i64),
    //Integer128(i128),
}

impl Integer {
    fn get_string_value(&self) -> String {
        match self {
            &Integer::Integer32(val) => val.to_string(),
            &Integer::Integer64(val) => val.to_string(),
        }
    }
}

#[derive(Clone)]
pub struct Lexer {
    input: String,
    position: i32,
    read_position: i32,
    ch: char,
}

impl Lexer {
    pub fn new(input: String) -> Lexer {
        let mut l = Lexer {
            input: input,
            position: 0,
            read_position: 0,
            ch: '0',
        };

        l.read_char();
        l
    }

    fn read_char(&mut self) {
        if self.read_position >= self.input.len() as i32 {
            self.ch = '\0';
        } else {
            self.ch = self.input.as_bytes()[self.read_position as usize] as char;
        }

        self.position = self.read_position;
        self.read_position += 1;
    }

    fn read_identifier(&mut self) -> Option<&str> {
        let pos = self.position as usize;

        //todo exercise
        while self.ch.is_alphabetic() {
            self.read_char();
        }

        if let Some(res) = self.input.get(pos..self.position as usize) {
            Some(res)
        } else {
            None
        }
    }

    fn skip_whitespace(&mut self) {
        while self.ch == ' ' || self.ch == '\t' || self.ch == '\n' || self.ch == '\r' {
            self.read_char();
        }
    }

    fn read_digit(&mut self) -> Option<&str> {
        let pos = self.position as usize;

        while self.ch.is_digit(10) {
            self.read_char();
        }

        if let Some(res) = self.input.get(pos..self.position as usize) {
            Some(res)
        } else {
            None
        }
    }
    fn peek_char(&mut self) -> Option<char> {
        if self.read_position >= self.input.len() as i32 {
            None
        } else {
            Some(self.input.as_bytes()[self.read_position as usize] as char)
        }
    }

    pub fn next_token(&mut self) -> Option<Token> {
        //if self.position + 1 == self.input.len() as i32 {
        //   return Some(Token::EOF);
        //}
        let mut tok: Token = Token::ILLEGAL;
        self.skip_whitespace();

        if self.ch == '\0' {
            return Some(Token::EOF);
        }
        match self.ch {
            '=' => {
                if let Some(val) = self.peek_char() {
                    if val == '=' {
                        self.read_char();
                        self.read_char();
                        tok = Token::EQ;
                        return Some(tok);
                    } else {
                        tok = Token::ASSIGN;
                    }
                }
            }
            '+' => tok = Token::PLUS,
            '-' => tok = Token::MINUS,
            '!' => {
                if let Some(val) = self.peek_char() {
                    if val == '=' {
                        self.read_char();
                        self.read_char();
                        tok = Token::NOT_EQ;
                        return Some(tok);
                    } else {
                        tok = Token::BANG;
                    }
                }
            }
            '/' => tok = Token::SLASH,
            '*' => tok = Token::ASTERISK,
            '<' => tok = Token::LT,
            '>' => tok = Token::GT,
            ';' => tok = Token::SEMICOLON,
            '(' => tok = Token::LPAREN,
            ')' => tok = Token::RPAREN,
            ',' => tok = Token::COMMA,
            '{' => tok = Token::LBRACE,
            '}' => tok = Token::RBRACE,
            _ => {
                //let start = self.position;

                if self.ch.is_alphabetic() {
                    if let Some(lit) = self.read_identifier() {
                        tok = lookup_ident(lit);
                        return Some(tok);
                    }
                } else if self.ch.is_digit(10) {
                    if let Some(lit_int) = self.read_digit() {
                        tok = Token::INT(Integer::Integer32(lit_int.parse().expect("must be int")));
                        return Some(tok);
                    }
                } else {
                    tok = Token::ILLEGAL
                }
            }
        }

        self.read_char();

        return Some(tok);
    }
}

static KEYWORDS: [(&str, Token); 7] = [
    ("let", Token::LET),
    ("fn", Token::FUNCTION),
    ("true", Token::TRUE),
    ("false", Token::FALSE),
    ("if", Token::IF),
    ("else", Token::ELSE),
    ("return", Token::RETURN),
];

fn lookup_ident(ident: &str) -> Token {
    for key in &KEYWORDS {
        if key.0 == ident {
            return key.1.clone();
        }
    }
    return Token::IDENT(String::from(ident));
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lexer() {
        let input = r#"
            let five = 5;
            let ten = 10;

            let add = fn(x, y) {
                x + y;
            };

            let result = add(five, ten);
            !-/*5;
            5 < 10 > 5;

            if (5 < 10) {
                return true;
            } else {
                return false;
            }
    
            10 == 10;
            10 != 9;
        "#;
        let tests: [(Token, &str); _] = [
            (Token::LET, "let"),
            (Token::IDENT("five".to_string()), "five"),
            (Token::ASSIGN, "="),
            (Token::INT(Integer::Integer32(5)), "5"),
            (Token::SEMICOLON, ";"),
            (Token::LET, "let"),
            (Token::IDENT("ten".to_string()), "ten"),
            (Token::ASSIGN, "="),
            (Token::INT(Integer::Integer32(10)), "10"),
            (Token::SEMICOLON, ";"),
            (Token::LET, "let"),
            (Token::IDENT("add".to_string()), "add"),
            (Token::ASSIGN, "="),
            (Token::FUNCTION, "fn"),
            (Token::LPAREN, "("),
            (Token::IDENT(String::from("x")), "x"),
            (Token::COMMA, ","),
            (Token::IDENT(String::from("y")), "y"),
            (Token::RPAREN, ")"),
            (Token::LBRACE, "{"),
            (Token::IDENT(String::from("x")), "x"),
            (Token::PLUS, "+"),
            (Token::IDENT(String::from("y")), "y"),
            (Token::SEMICOLON, ";"),
            (Token::RBRACE, "}"),
            (Token::SEMICOLON, ";"),
            (Token::LET, "let"),
            (Token::IDENT(String::from("result")), "result"),
            (Token::ASSIGN, "="),
            (Token::IDENT(String::from("add")), "add"),
            (Token::LPAREN, "("),
            (Token::IDENT(String::from("five")), "five"),
            (Token::COMMA, ","),
            (Token::IDENT(String::from("ten")), "ten"),
            (Token::RPAREN, ")"),
            (Token::SEMICOLON, ";"),
            (Token::BANG, "!"),
            (Token::MINUS, "-"),
            (Token::SLASH, "/"),
            (Token::ASTERISK, "*"),
            (Token::INT(Integer::Integer32(5)), "5"),
            (Token::SEMICOLON, ";"),
            (Token::INT(Integer::Integer32(5)), "5"),
            (Token::LT, "<"),
            (Token::INT(Integer::Integer32(10)), "10"),
            (Token::GT, ">"),
            (Token::INT(Integer::Integer32(5)), "5"),
            (Token::SEMICOLON, ";"),
            (Token::IF, "if"),
            (Token::LPAREN, "("),
            (Token::INT(Integer::Integer32(5)), "5"),
            (Token::LT, "<"),
            (Token::INT(Integer::Integer32(10)), "10"),
            (Token::RPAREN, ")"),
            (Token::LBRACE, "{"),
            (Token::RETURN, "return"),
            (Token::TRUE, "true"),
            (Token::SEMICOLON, ";"),
            (Token::RBRACE, "}"),
            (Token::ELSE, "else"),
            (Token::LBRACE, "{"),
            (Token::RETURN, "return"),
            (Token::FALSE, "false"),
            (Token::SEMICOLON, ";"),
            (Token::RBRACE, "}"),
            (Token::INT(Integer::Integer32(10)), "10"),
            (Token::EQ, "=="),
            (Token::INT(Integer::Integer32(10)), "10"),
            (Token::SEMICOLON, ";"),
            (Token::INT(Integer::Integer32(10)), "10"),
            (Token::NOT_EQ, "!="),
            (Token::INT(Integer::Integer32(9)), "9"),
            (Token::SEMICOLON, ";"),
        ];

        let mut l = Lexer::new(input.to_string());
        for test in tests {
            if let Some(tok) = l.next_token() {
                assert_eq!(tok, test.0);
                println!("{:?} {:?}", tok.literal(), test.0.literal());
            }
        }
    }
}
